// booking.proto
syntax = "proto3";
package booking;

option java_package = "com.example.booking";
option go_package = "github.com/example/distributed-ticket-booking/proto/booking;booking";

service BookingService {
  // Admin RPC: Add a new show or update seat capacity/price for an existing show.
  rpc AddShow(AddShowRequest) returns (AddShowResponse);
  
  // NEW: List all available shows
  rpc ListShows(ListShowsRequest) returns (ListShowsResponse);
  
  // Book a specific seat for a user (goes through Payment -> Raft).
  rpc BookSeat(BookRequest) returns (BookResponse);

  // Query seat availability.
  rpc QuerySeat(QueryRequest) returns (QueryResponse);

  // List seat status for a show / event (paged).
  rpc ListSeats(ListSeatsRequest) returns (ListSeatsResponse);
}

// ===== NEW MESSAGES FOR LISTSHOWS =====
message ListShowsRequest {
  // No parameters needed - returns all shows
}

message ShowInfo {
  string show_id = 1;
  int32 total_seats = 2;
  int64 price_cents = 3;
  int32 available_seats = 4;
  int32 booked_seats = 5;
}

message ListShowsResponse {
  repeated ShowInfo shows = 1;
}
// ======================================

message AddShowRequest {
  string user_id = 1;         // User's session token (must be Admin)
  string show_id = 2;         // The show/event ID
  int32 total_seats = 3;      // Total number of seats available
  int64 price_cents = 4;      // Price of a single seat in cents
}

message AddShowResponse {
  bool success = 1;
  string message = 2;
}

message BookRequest {
  string user_id = 1;
  int32 seat_id = 2;
  string show_id = 3;
  string correlation_id = 4;
  string payment_token = 5;
}

message BookResponse {
  bool success = 1;
  string message = 2;
  string booking_id = 3;
  Seat seat = 4;
}

message QueryRequest {
  string show_id = 1;
  int32 seat_id = 2;
}

message QueryResponse {
  bool available = 1;
  Seat seat = 2;
}

message ListSeatsRequest {
  string show_id = 1;
  int32 page_size = 2;
  int32 page_token = 3;
}

message ListSeatsResponse {
  repeated Seat seats = 1;
  int32 next_page_token = 2;
}

message Seat {
  int32 seat_id = 1;
  string show_id = 2;
  bool reserved = 3;
  string reserved_by = 4;
  int64 reserved_at = 5;
  string booking_id = 6;
  int64 price_cents = 7;
}